proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -368,6 +368,1382 @@ lemma invoke_sched_context_corres:
      reused in Finalise and IpcCancel *)
   sorry
 
+(* FIXME RT: move *)
+lemmas update_sched_context_decompose_ext
+  = update_sched_context_decompose[where f="f x" and g="g y" for f g x y]
+lemmas update_sched_context_decompose2
+  = update_sched_context_decompose[where g="\<lambda>sc. g (h sc)" for g h]
+lemmas update_sched_context_decompose_ext2
+  = update_sched_context_decompose2[where f="f x" and g="g y" for f g x y]
+
+lemma refill_new_bundled:
+   "refill_new sc_ptr max_refills budget period
+    = (do cur_time \<leftarrow> gets cur_time;
+          set_sc_obj_ref sc_period_update sc_ptr period;
+          set_sc_obj_ref sc_budget_update sc_ptr budget;
+          update_sched_context sc_ptr (\<lambda>sc. sc\<lparr>sc_refill_max := max_refills,
+                                               sc_refills := [\<lparr>r_time = cur_time, r_amount = budget\<rparr>]\<rparr>);
+          maybe_add_empty_tail sc_ptr
+       od)"
```

Comment:
Sure. The C will update these fields individually. So I thought it would be best for the Haskell to do so too, rather than deal with worse rewrites in CRefine. However, the refill list (with `sc_refill_max`) in the abstract corresponds to many fields in the Haskell: there's `scRefillCount` which gives the actual number of refills we'd like to extract from the buffer, the `scRefillHead` which tells you where to start in the circular buffer, `scRefillMax` which tells you where to wrap around, and `scRefills` (which I'd almost like to be renamed to `scRefillBuffer` or something) which is the actual circular buffer. So the updates are quite asymmetric, and the bundled update allows the `setSchedContext_corres` rules to be nicely applied.

Once upon a time in a kernel long long ago, we had `refill_new` and `refill_update` written with bundled updates. But Matt changed them to the individual updates when the heap projections were introduced. Had the aspec and Haskell both used bundled updates, it would have been much more straightforward to prove the corres rules. and we could have just done some rewriting of the Haskell functions for CRefine. But with the way things are now, I thought this was the best option. 

In hindsight, it probably warranted a discussion, but a lot of this work was done a long time ago as a "task on the side"...